sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/datagen/DataGeneratorTableProviderTest.java

```diff
@@ -0,0 +1,296 @@
+            + "  event_id BIGINT,\n"
+            + "  event_name VARCHAR,\n"
+            + "  score DOUBLE\n"
+            + ") TYPE 'datagen' TBLPROPERTIES '{\n"
+            + "  \"number-of-rows\": \"10\",\n"
+            + "  \"fields.event_id.kind\": \"sequence\",\n"
+            + "  \"fields.event_id.start\": \"100\",\n"
+            + "  \"fields.event_id.end\": \"109\",\n"
+            + "  \"fields.event_name.kind\": \"random\",\n"
+            + "  \"fields.event_name.length\": \"15\",\n"
+            + "  \"fields.score.kind\": \"random\",\n"
+            + "  \"fields.score.min\": \"50.0\",\n"
+            + "  \"fields.score.max\": \"100.0\"\n"
+            + "}'";
+
+    PCollection<Row> result =
+        pipeline.apply(
+            "testFieldGenerators",
+            SqlTransform.query("SELECT * FROM complex_table").withDdlString(createDdl));
+
+    result.apply("ValidateFields", ParDo.of(new ValidateFieldsFn(100, 109, 15, 50.0, 100.0)));
+    pipeline.run().waitUntilFinish();
+  }
+
+  private static class ValidateNullsFn extends DoFn<Row, Void> {
+    @ProcessElement
+    public void processElement(@Element Row row) {
+      Assert.assertNull("Field should be null", row.getValue("nullable_field"));
+      Assert.assertNotNull("Field should not be null", row.getInt64("non_nullable_field"));
+    }
+  }
+
+  @Test
+  public void testNullRate() {
+    String createDdl =
+        "CREATE EXTERNAL TABLE null_rate_table (\n"
+            + "  nullable_field VARCHAR,\n"
+            + "  non_nullable_field BIGINT\n"
+            + ") TYPE 'datagen' TBLPROPERTIES '{\n"
+            + "  \"number-of-rows\": \"50\",\n"
+            + "  \"fields.nullable_field.null-rate\": \"1.0\",\n"
+            + "  \"fields.non_nullable_field.kind\": \"sequence\"\n"
+            + "}'";
+
+    PCollection<Row> result =
+        pipeline.apply(
+            "testNullRate",
+            SqlTransform.query("SELECT * FROM null_rate_table").withDdlString(createDdl));
+
+    result.apply("ValidateNulls", ParDo.of(new ValidateNullsFn()));
+
+    pipeline.run().waitUntilFinish();
+  }
+
+  private static class ValidateAllTypesFn extends DoFn<Row, Void> {
+    @ProcessElement
+    public void processElement(@Element Row row) {
+      Assert.assertNotNull(row.getBoolean("is_active"));
+
+      BigDecimal cost = row.getDecimal("cost");
+      Assert.assertTrue("Cost should be >= 10.50", cost.compareTo(BigDecimal.valueOf(10.50)) >= 0);
+      Assert.assertTrue("Cost should be <= 99.99", cost.compareTo(BigDecimal.valueOf(99.99)) <= 0);
+
+      Instant now = Instant.now();
+      Instant pastTimestamp = row.getDateTime("past_timestamp").toInstant();
+      Instant nowTimestamp = row.getDateTime("now_timestamp").toInstant();
+
+      Assert.assertTrue("past_timestamp should be in the past", pastTimestamp.isBefore(now));
+      Assert.assertTrue(
+          "now_timestamp should be very recent",
+          now.plus(Duration.millis(100)).isAfter(nowTimestamp));
```

Comment:
Wouldn't this always be true because `now` is always freshly created? Maybe something like this works:

```suggestion
          now.minus(Duration.millis(100)).isBefore(nowTimestamp));
```